python/cudf/cudf/core/frame.py

```diff
@@ -97,6 +97,133 @@ def _as_column(self):
 
         return self._data[None].copy(deep=False)
 
+    def _scatter(self, key, value):
+        result = self._from_table(libcudfxx.copying.scatter(value, key, self))
+
+        result._copy_categories(self)
+        return result
+
+    def _empty_like(self):
+        result = self._from_table(libcudfxx.copying.table_empty_like(self))
+
+        result._copy_categories(self)
+        return result
+
+    def _slice(self, arg):
+        """
+       _slice : slice the frame as per the arg
+
+       Parameters
+       ----------
+       arg : should always be of type slice and doesn't handle step
+
+       """
+        num_rows = len(self)
+        if num_rows == 0:
+            return self
+        start, stop, stride = arg.indices(num_rows)
+        if start < 0:
+            start = start + num_rows
+        if stop < 0:
+            stop = stop + num_rows
+
+        if start > stop:
+            return self._empty_like()
+        else:
+            start = len(self) if start > num_rows else start
+            stop = len(self) if stop > num_rows else stop
+
+            result = self._from_table(
+                libcudfxx.copying.table_slice(self, [start, stop])[0]
+            )
+
+            result._copy_categories(self)
+            return result
+
+    def _normalize_columns_and_scalars_type(self, other):
```

Comment:
Can you add a docstring for this function? It does a lot under the hood where something explaining its behavior would be helpful.